src/cargo/util/lints.rs

```diff
@@ -976,6 +976,20 @@ pub fn unexpected_target_cfgs(
         return Ok(());
     };
 
+    // If we have extra `--check-cfg` args comming from the lints config, we need to
+    // refetch the `--print=check-cfg` with those extra args.
+    let lint_rustflags = pkg.manifest().lint_rustflags();
+    let check_cfg = if lint_rustflags.iter().any(|a| a == "--check-cfg") {
+        Some(target_info.check_cfg_with_extra_args(gctx, &rustc, lint_rustflags)?)
```

Comment:
Is there a reason we are doing both a "global" and " extra_args" calls rather than just the latter?

Reply:
The "global" call is always done, cached and already contains the `--print=check-cfg`, so getting it is mostly "free", and even if we don't use the print-check-cfg it parsed it is still useful as it gives us all the `RUSTFLAGS` args that we can just re-use for free to the "extra_args" case.

Reply:
If we don't need to ask for it from rustc twice, we probably shouldn't. Removing the global check_cfg seems like  it would simplify this overall PR as we no longer need to interject ourselves into the regular rustc query, even if we still need it for processing rustflags.

Reply:
Well we don't really ask for rustc twice, we are just re-using a already available information for when there is no `[lints.rust.unexpected_cfgs.check-cfg]` specified, which I think is the vast majority of projects.

The same "global" call will be used for cfgs, split-debuginfo support, ...

-----

Schematically, if we have 3 packages, we would currently do 1 `rustc` call for the "global".

Now imaging that one the package as extra check-cfg args from the lint config, with the current strategy we would only do 2 `rustc` calls, the "global" and the one specialized for the package.

If we were to not "interject ourselves into the regular rustc query" (aka "global") there would 4 `rustc` calls (the global and 1 per package) instead of 2.

-----

Btw I'm getting mixed signal with this comment and the one this [conversation](https://github.com/rust-lang/cargo/pull/14581#discussion_r1775906095):
> Calling rustc is "expensive" for no-op builds which is especially important for cargo-script. Yes, this is cached, but I'd like to better understand why we need this **over adding it to another call**.

Which is why I changed strategy to reduce the number of calls.